#  FARS Analysis

 ## Written By Damilola OIaleye

### Problem 1
For each state, what day of the week has the most accidents? Use the DAY_WEEK column. Output the day and the count. For the values output the day name, where 1 is Sunday, 2 is Monday, ... and 7 is Saturday.

In [1]:
import os 
import pandas as pd
files=[f for f in os.listdir("FARS")]
filename=[f for f in os.listdir("FARS") if f.startswith("a")]
days={1:"Sunday",2:"Monday",3:"Teusday",4:"Wednessday",5:"Thursday",6:"Friday",7:"Saturday"}

states={1:"Alabama",2:"Alaska",4:"Arizona",5:"Arkansas",6:"California",8:"Colorado",9:"Connecticut",10:"Delaware",
31:"Nebraska",32:"Nevada",33:"New Hampshire",34:"New Jersey",35:"New Mexico",36:"New York",37:"North Carolina",
38:"North Dakota",11:"District of Columbia",39:"Ohio",12:"Florida",40:"Oklahoma",13:"Georgia",41:"Oregon",15:"Hawaii",
42:"Pennsylvania",16:"Idaho",43:"Puerto Rico",17:"Illinois",44:"Rhode Island",18:"Indiana",45:"South Carolina",
19:"Iowa",46:"South Dakota",20:"Kansas",47:"Tennessee",21:"Kentucky",48:"Texas",22:"Louisiana",49:"Utah",23:"Maine",
50:"Vermont",24:"Maryland",52:"Virgin Islands (since 2004)",25:"Massachusetts",51:"Virginia",26:"Michigan",53:"Washington",
27:"Minnesota",54:"West Virginia",28:"Mississippi",55:"Wisconsin",29:"Missouri",56:"Wyoming",30:"Montana"}

data=[]
for file in filename:
    pth=os.path.join("FARS",file)
    fil=pd.read_csv(pth)
    data.append(fil)
totalData=pd.concat(data)    
for i in range(len(data)):
    col=data[i].DAY_WEEK
    get_mode=col.mode().values[0]
    state=data[i].STATE[2]
    statename=states[state]
    day=days[get_mode]
    count=col.value_counts()[get_mode]
    print("For",statename,", ",day,"has :",count,"count(s)")

For Vermont ,  Wednessday has : 11 count(s)
For Rhode Island ,  Thursday has : 14 count(s)
For South Carolina ,  Saturday has : 182 count(s)
For Virginia ,  Saturday has : 139 count(s)
For Tennessee ,  Saturday has : 187 count(s)
For Washington ,  Saturday has : 87 count(s)
For South Dakota ,  Saturday has : 18 count(s)
For Pennsylvania ,  Friday has : 186 count(s)
For Wyoming ,  Thursday has : 19 count(s)
For Wisconsin ,  Saturday has : 112 count(s)
For Oregon ,  Saturday has : 76 count(s)
For Oklahoma ,  Saturday has : 109 count(s)
For West Virginia ,  Wednessday has : 41 count(s)
For New Hampshire ,  Friday has : 26 count(s)
For Minnesota ,  Saturday has : 58 count(s)
For Michigan ,  Saturday has : 175 count(s)
For Nevada ,  Saturday has : 56 count(s)
For Indiana ,  Friday has : 127 count(s)
For Maryland ,  Saturday has : 96 count(s)
For Montana ,  Saturday has : 34 count(s)
For Nebraska ,  Sunday has : 38 count(s)
For Massachusetts ,  Saturday has : 61 count(s)
For Iowa ,  Saturday

The files were read using OS library and then saved into read_data using pd.read_csv.The statecode was used to map the state to get its index, for loop was used to to get the index for the state and days of the week 

The mode function was used to get the highest count  from the "DAY_WEEK" column. 

### Problem 2
For whole United States, i.e., all the data, what day of the week has the most accidents? 
Output the day and the count.

In [2]:
dayweek=[]
for i in range(len(data)):
    col2=data[i].DAY_WEEK
    for i in col2:
        dayweek.append(i)
dayweek_lst=pd.Series(dayweek)
get_mode2=dayweek_lst.mode().values[0]
count2=dayweek_lst.value_counts()[get_mode2]

print(days[get_mode2],"has the most accident with ",count2,"counts")   

Saturday has the most accident with  6104 counts


An empty list was created to store the days of the week from the "DAY_WEEK" column. 
It was converted into a pandas SERIES and mode was used to get the highest values series and its count. 

Saturday has the most accident with  6104 counts

### Problem 3
For each state, what hour of the day has the most accidents? Output the hour and the count.

* A value of 99 in the HOUR means unknown.

In [3]:
len(data)

51

In [4]:
for j in range(len(data)):
    col3=data[j].HOUR 
    getmode3=col3.mode().values[0]
    state=data[j].STATE[2]
    statename=states[state]
    count3=col3.value_counts()[getmode3]
    mode=col3.mode()[0]
    print("For",statename,",", mode,"hour has the most accident with",count3,"count")

For Vermont , 15 hour has the most accident with 6 count
For Rhode Island , 1 hour has the most accident with 6 count
For South Carolina , 17 hour has the most accident with 64 count
For Virginia , 17 hour has the most accident with 60 count
For Tennessee , 16 hour has the most accident with 61 count
For Washington , 15 hour has the most accident with 34 count
For South Dakota , 14 hour has the most accident with 10 count
For Pennsylvania , 17 hour has the most accident with 64 count
For Wyoming , 12 hour has the most accident with 11 count
For Wisconsin , 17 hour has the most accident with 45 count
For Oregon , 17 hour has the most accident with 31 count
For Oklahoma , 16 hour has the most accident with 41 count
For West Virginia , 15 hour has the most accident with 24 count
For New Hampshire , 16 hour has the most accident with 11 count
For Minnesota , 15 hour has the most accident with 28 count
For Michigan , 15 hour has the most accident with 59 count
For Nevada , 23 hour has the m

Hour and state value were stored in col3 and state respectively and mode was used to generate the maximum value count


### Problem 4
For whole United States, what hour of the day has the most accidents? Output the hour and the count.

In [5]:
hours=[]
for i in range(len(data)):
    col4=data[i].HOUR
    #mode=col.mode().values[0]
    for i in col4:
        hours.append(i)
dayweek4=pd.Series(hours)
getmode4=dayweek4.mode().values[0]
count4=dayweek4.value_counts()[getmode4]
print("Highest Hours:",col4[getmode4],"With Count:",count4)

Highest Hours: 19 With Count: 1984


An empty was craeted to store each. hour, and then generated the mode for both hour and count. A result of Highest Hours: 19 and Count: 1984 was generated meaning 19 th hour has the most accident 

### Problem 5
For each state, what is the percentage of fatal accidents involved at least one drunk driver? If the column, DRUNK_DR, has a 0, then no drunk drivers were involved. Any number larger than 0 indicates the number of drunk drivers involved in the accident.

In [6]:
for j in range(len(data)):
    mask=data[j].DRUNK_DR !=0
    drunkAcc = data[j][mask]
    fatal=data[j].FATALS
    sum_count_drunk = drunkAcc.DRUNK_DR.count()
    sum_count_accident = sum(fatal.value_counts())

    percent = (sum_count_drunk/sum_count_accident)*100
    state=data[j].STATE[2]
    statename=states[state]
    
    
    print("The percentage for ", statename,"is:")
    print(percent, "%")  

The percentage for  Vermont is:
47.368421052631575 %
The percentage for  Rhode Island is:
37.5 %
The percentage for  South Carolina is:
35.47008547008547 %
The percentage for  Virginia is:
29.224376731301938 %
The percentage for  Tennessee is:
22.56728778467909 %
The percentage for  Washington is:
31.547619047619047 %
The percentage for  South Dakota is:
41.74757281553398 %
The percentage for  Pennsylvania is:
23.621323529411764 %
The percentage for  Wyoming is:
28.999999999999996 %
The percentage for  Wisconsin is:
31.985294117647058 %
The percentage for  Oregon is:
26.905829596412556 %
The percentage for  Oklahoma is:
27.564102564102566 %
The percentage for  West Virginia is:
27.6 %
The percentage for  New Hampshire is:
30.0 %
The percentage for  Minnesota is:
27.73109243697479 %
The percentage for  Michigan is:
24.79591836734694 %
The percentage for  Nevada is:
29.7029702970297 %
The percentage for  Indiana is:
17.708333333333336 %
The percentage for  Maryland is:
23.728813559322035

I obtained drunk driver element greater than 0 from the "DRUNK_DRIVER" column. the sum total of the drunk drivers was calculated as well as the percentage using percent = (sum_count_drunk/sum_count_accident)*100

### Problem 6
For whole United States, what is the percentage of fatal accidents involved at least one drunk driver?


In [7]:
drunkDr = []
fatalAccident = []
for j in range(len(data)):
    mask=data[j].DRUNK_DR !=0
    drunkAcc = data[j][mask]
    fatal=data[j].FATALS
    drunkSum = drunkAcc.DRUNK_DR.count()
    accidentSum = sum(fatal.value_counts())
    
    drunkDr.append(drunkSum)
    fatalAccident.append(accidentSum)

accident_percent = (sum(drunkDr)/sum(fatalAccident)*100)
print("The percentage of fatal accident for the whole state is:" , accident_percent,"%")

The percentage of fatal accident for the whole state is: 25.320131246551874 %


!=0 was used to obtain the value of drunk driver not equal to 0,drunkDr contains the sum of drunk driver greater than 0 and fatalAccident contains the sum of all fatal accident.

accident_percent = (sum(drunkDr)/sum(fatalAccident)*100) wasn used to calculate the percentage

25.320131246551874 % was obtained 

### Problem 7
For whole United States, how many fatalities were caused by each type of collision? Use the MAN_COLL column. The values in the column are below.

* 0	Not Collision with Motor Vehicle in Transport
* 1	Front-to-Rear
* 2	Front-to-Front
* 6	Angle
* 7	Sideswipe – Same Direction
* 8	Sideswipe – Opposite Direction
* 9	Rear-to-Side
* 10 Rear-to-Rear
* 11 Other (End-Swipes and Others)
* 98 Not Reported
* 99 Unknown

In [8]:
colision_count = totalData["MAN_COLL"].value_counts()
print("Not Collision with Motor Vehicle in Transport caused", colision_count[0],"fatalities")
print("Front-to-Rear caused", colision_count[1],"fatalities")
print("Front-to-Front caused", colision_count[2],"fatalities")
print("Angle caused ", colision_count[6],"fatalities")
print("Sideswipe – Same Direction caused", colision_count[7],"fatalities")
print("Sideswipe – Opposite Direction caused", colision_count[8],"fatalities")
print("Rear-to-Side caused", colision_count[9],"fatalities")
print("Rear-to-Rear caused", colision_count[10],"fatalities")
print("Other (End-Swipes and Others) caused", colision_count[11],"fatalities")
print("Not Reported caused", colision_count[98],"fatalities")
print("Unknown caused", colision_count[99],"fatalities")

Not Collision with Motor Vehicle in Transport caused 21296 fatalities
Front-to-Rear caused 2350 fatalities
Front-to-Front caused 3511 fatalities
Angle caused  6122 fatalities
Sideswipe – Same Direction caused 519 fatalities
Sideswipe – Opposite Direction caused 421 fatalities
Rear-to-Side caused 32 fatalities
Rear-to-Rear caused 2 fatalities
Other (End-Swipes and Others) caused 86 fatalities
Not Reported caused 23 fatalities
Unknown caused 77 fatalities


colision_count was used to store the count of collision value from "MAN_COLL" column. 

Each count value was used to map the code for the collision type

### Problem 8
For each state, what is its fatal accident rate per 10,000 people? To calculate this, count the number of accidents in a state, divide by the state's 2016 population estimate from the nst-est2017-alldata.csv Census data file, and then multiply by 10000. Output the states' rates in order from highest to lowest.

In [9]:
files=[f for f in os.listdir("FARS")]
filename2=[f for f in os.listdir("FARS") if f.startswith("n")]
PopLst = []
for file in files:
    if file.startswith("n"):
        pathname = os.path.join("FARS", file)
        frame = pd.read_csv(pathname)
        popestimate = frame["POPESTIMATE2016"]
        b = popestimate[5:57]
for item in b:  
    PopLst.append(item)
stateFatal_dict={}
print(len(data))
for j in range(len(data)):  
    fatal=data[j].FATALS
    sum_count_accident =sum(fatal.value_counts())
    fatal_state_acc = (sum_count_accident/PopLst[j])*10000
    state=data[j].STATE[2]
    statename=states[state]
    stateFatal_dict[statename] = fatal_state_acc

from operator import itemgetter
sortedlist = sorted(stateFatal_dict.items(), key = itemgetter(1), reverse = True)
sortedlist


51


[('California', 44.43132666620784),
 ('Alabama', 15.031587188018365),
 ('Pennsylvania', 11.420198216013889),
 ('New York', 7.228383201686872),
 ('Texas', 5.901695797808976),
 ('Colorado', 5.372327315299772),
 ('Utah', 4.428031662990887),
 ('Oklahoma', 4.367658885841016),
 ('Michigan', 3.1301213816355777),
 ('New Jersey', 2.982800928704767),
 ('Florida', 2.887752101103027),
 ('Ohio', 2.6853976339428143),
 ('Virginia', 2.416145204303148),
 ('Arizona', 1.744014200509615),
 ('Indiana', 1.7312453492505713),
 ('West Virginia', 1.4880722084062985),
 ('Wyoming', 1.461270487012228),
 ('Kentucky', 1.3809829775296152),
 ('South Carolina', 1.3548248700685317),
 ('Kansas', 1.2962472790714923),
 ('Montana', 1.2854900498559652),
 ('Idaho', 1.2687045050493893),
 ('Georgia', 1.2234832378494436),
 ('New Mexico', 1.1991632654086617),
 ('Arkansas', 1.1943252906456674),
 ('Illinois', 1.1920070165597465),
 ('North Dakota', 1.1839237088847672),
 ('Delaware', 1.0968582575461012),
 ('Nevada', 1.042049625635934

I used the os to load the data, and then created an empty stateFatal_dict to store fatal accident rate per 10,000 people  using the formular fatal_state_acc = (sum_count_accident/PopLst[j])*10000 and then finally used item getter to get the key value from the dic created

### Problem 9
For each state, what is the rate of fatal accidents caused by drunk driving per 10,000 people? To calculate this, count the number of accidents in which a drunk driver was involved, divide by the state's 2016 population estimate from the nst-est2017-alldata.csv Census data file, and then multiply by 10000. Output the states' rates in order from highest to lowest.

In [10]:
files=[f for f in os.listdir("FARS")]
filename2=[f for f in os.listdir("FARS") if f.startswith("n")]
PopLst2 = []
for file in files:
    if file.startswith("n"):
        pathname = os.path.join("FARS", file)
        frame = pd.read_csv(pathname)
        popestimate = frame["POPESTIMATE2016"]
        b = popestimate[5:57]
for item in b:  
    PopLst2.append(item)
stateFatal_dict={}
for j in range(len(data)): 
    mask=data[j].DRUNK_DR !=0
    drunkAcc = data[j][mask]
    sum_count_drunk = drunkAcc.DRUNK_DR.count()
                
    fatal_state_acc = (sum_count_drunk/PopLst2[j])*10000
    state=data[j].STATE[2]
    statename=states[state]
    stateFatal_dict[statename] = fatal_state_acc

from operator import itemgetter
sortedlist = sorted(stateFatal_dict.items(), key = itemgetter(1), reverse = True)
sortedlist

[('California', 10.41628063339457),
 ('Pennsylvania', 2.697601968304751),
 ('Alabama', 2.1977880947262713),
 ('Colorado', 1.8100314252264464),
 ('Texas', 1.4585347407558433),
 ('New York', 1.243431721740954),
 ('Oklahoma', 1.203905974943357),
 ('Ohio', 0.9002330149874771),
 ('Utah', 0.8719290147202134),
 ('Michigan', 0.77614234258923),
 ('Virginia', 0.7061033768808369),
 ('New Jersey', 0.6552726117541228),
 ('Montana', 0.6164338250771294),
 ('Florida', 0.6133888711173494),
 ('North Dakota', 0.568747664072094),
 ('South Carolina', 0.4805575393832826),
 ('Wyoming', 0.4237684412335461),
 ('West Virginia', 0.41070792952013835),
 ('Arizona', 0.40727268035022224),
 ('Idaho', 0.3609245574709469),
 ('New Mexico', 0.34836027821927606),
 ('Kentucky', 0.34388829060370496),
 ('Delaware', 0.3404042868246521),
 ('Illinois', 0.314937048243602),
 ('Nevada', 0.3095196907829507),
 ('Indiana', 0.30657469726312203),
 ('Washington', 0.2875171447920067),
 ('Arkansas', 0.279002219536078),
 ('Georgia', 0.2658

I used the os to load the data, and then created an empty stateFatal_dict to store fatal accident rate per 10,000 people. 
I used data[j].DRUNK_DR !=0 to get the value o
using the formular fatal_state_acc = (sum_count_accident/PopLst[j])*10000 and then finally used item getter to get the key value from the dic created

In [11]:
files=[f for f in os.listdir("FARS")]
#filename2=[f for f in os.listdir("FARS") if f.startswith("n")]
PopLst2 = []
for file in files:
    if file.startswith("n"):
        pathname = os.path.join("FARS", file)
        frame = pd.read_csv(pathname)
        popestimate = frame["POPESTIMATE2016"]
        b = popestimate[5:57]
for item in b:  
    PopLst2.append(item)
stateFatal_dict={}
for j in range(len(data)): 
    mask=data[j].DRUNK_DR
    mask1=[]
    for w in mask:
        if w>0:
            mask1.append(w)
    #drunkAcc = data[j][mask]
    #sum_count_drunk = drunkAcc.DRUNK_DR.count()
                
    fatal_state_acc = (len(mask1)/PopLst2[j])*10000
    state=data[j].STATE[2]
    statename=states[state]
    stateFatal_dict[statename] = fatal_state_acc

from operator import itemgetter
sortedlist = sorted(stateFatal_dict.items(), key = itemgetter(1), reverse = True)
sortedlist

[('California', 10.41628063339457),
 ('Pennsylvania', 2.697601968304751),
 ('Alabama', 2.1977880947262713),
 ('Colorado', 1.8100314252264464),
 ('Texas', 1.4585347407558433),
 ('New York', 1.243431721740954),
 ('Oklahoma', 1.203905974943357),
 ('Ohio', 0.9002330149874771),
 ('Utah', 0.8719290147202134),
 ('Michigan', 0.77614234258923),
 ('Virginia', 0.7061033768808369),
 ('New Jersey', 0.6552726117541228),
 ('Montana', 0.6164338250771294),
 ('Florida', 0.6133888711173494),
 ('North Dakota', 0.568747664072094),
 ('South Carolina', 0.4805575393832826),
 ('Wyoming', 0.4237684412335461),
 ('West Virginia', 0.41070792952013835),
 ('Arizona', 0.40727268035022224),
 ('Idaho', 0.3609245574709469),
 ('New Mexico', 0.34836027821927606),
 ('Kentucky', 0.34388829060370496),
 ('Delaware', 0.3404042868246521),
 ('Illinois', 0.314937048243602),
 ('Nevada', 0.3095196907829507),
 ('Indiana', 0.30657469726312203),
 ('Washington', 0.2875171447920067),
 ('Arkansas', 0.279002219536078),
 ('Georgia', 0.2658